pkg/gardenlet/controller/managedseed/managedseed_reconciler.go

```diff
@@ -0,0 +1,280 @@
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// reconciler implements the reconcile.Reconcile interface for ManagedSeed reconciliation.
+type reconciler struct {
+	ctx         context.Context
+	clientMap   clientmap.ClientMap
+	config      *config.GardenletConfiguration
+	imageVector imagevector.ImageVector
+	recorder    record.EventRecorder
+	logger      *logrus.Logger
+}
+
+// newReconciler returns the new ManagedSeed reconciler.
+func newReconciler(ctx context.Context, clientMap clientmap.ClientMap, config *config.GardenletConfiguration, imageVector imagevector.ImageVector, recorder record.EventRecorder, logger *logrus.Logger) reconcile.Reconciler {
+	return &reconciler{
+		ctx:         ctx,
+		clientMap:   clientMap,
+		config:      config,
+		imageVector: imageVector,
+		recorder:    recorder,
+		logger:      logger,
+	}
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	gardenClient, err := r.clientMap.GetClient(r.ctx, keys.ForGarden())
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get garden client: %w", err)
+	}
+
+	managedSeed := &seedmanagementv1alpha1.ManagedSeed{}
+	if err := gardenClient.Client().Get(r.ctx, request.NamespacedName, managedSeed); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.logger.Debugf("Skipping ManagedSeed %s because it has been deleted", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		r.logger.Errorf("Could not get ManagedSeed %s from store: %v", request.NamespacedName, err)
+		return reconcile.Result{}, err
+	}
+
+	if managedSeed.Spec.Shoot == nil || managedSeed.Spec.Shoot.Name == "" {
+		r.logger.Errorf("Skipping ManagedSeed %s because it does not specify a shoot", request.NamespacedName)
+		return reconcile.Result{}, nil
+	}
+
+	if managedSeed.DeletionTimestamp != nil {
+		return r.delete(gardenClient, managedSeed)
+	}
+	return r.createOrUpdate(gardenClient, managedSeed)
+}
+
+func (r *reconciler) createOrUpdate(gardenClient kubernetes.Interface, managedSeed *seedmanagementv1alpha1.ManagedSeed) (reconcile.Result, error) {
```

Comment:
Usually, we call this 
```suggestion
func (r *reconciler) reconcile(gardenClient kubernetes.Interface, managedSeed *seedmanagementv1alpha1.ManagedSeed) (reconcile.Result, error) {
```

Reply:
I know, but I find this naming not really accurate and was bothered by it, so I am trying to propose a better one. Both `createOrUpdate` and `delete` perform a reconciliation, just of different types of changes, so calling one of them `reconcile` is not really good naming. I am already avoiding it in other contexts, see e.g. https://github.com/gardener/remedy-controller/blob/master/pkg/controller/controller.go. I am aware that `reconcile` is used in this sense in a lot of places and if consistency is considered more important, it should be used here as well.

Reply:
I understand your concerns and appreciate the proposal, but I would stay with the current names for now, otherwise this makes it harder to understand (especially for new colleagues) why it suddenly is called `CreateOrUpdate` instead of `Reconcile` here.

Reply:
Sure, will use `reconcile` for consistency in this PR. What is the way forward to improve this? Should I simply open a new PR with a naming change (`Reconcile` => `CreateOrUpdate`, of course only where appropriate) in g/g (extension controllers having to adapt on revendor)?

Reply:
OK, sounds good. Let's chat about this in our next weekly meeting and hear some other opinions. If we agree to change to `CreateOrUpdate` then we can follow this up as you suggested. Yes, the extensions could adapt accordingly after the next revendoring.

Reply:
OK, thanks